libopae/plugins/xfpga/sysfs.c

```diff
@@ -215,74 +222,152 @@ STATIC sysfs_fpga_region *make_region(sysfs_fpga_device *device, char *name,
+	ASSERT_NOT_NULL(type);
+	ASSERT_NOT_NULL(num);
+	reg_res = regcomp(&re, fmt, REG_EXTENDED);
+	if (reg_res) {
+		regerror(reg_res, &re, err, sizeof(err));
+		FPGA_MSG("Error compiling regex: %s", err);
+		return FPGA_EXCEPTION;
+	}
+	reg_res = regexec(&re, inpstr, RE_REGION_GROUPS, matches, 0);
+	if (!reg_res) {
+		ptr = inpstr + matches[RE_REGION_GROUP_TYPE].rm_so;
+		end = inpstr + matches[RE_REGION_GROUP_TYPE].rm_eo;
+		if (strncpy_s(type, type_len, ptr, end-ptr)) {
+			FPGA_MSG("Error copying type from string: %s", inpstr);
+			goto out_free;
 
-		reg_res = regcomp(&re, SYSFS_FORMAT(sysfs_region_fmt), REG_EXTENDED);
-		if (reg_res) {
-			regerror(reg_res, &re, err, 128);
-			FPGA_MSG("Error compiling regex: %s", err);
-			return FPGA_EXCEPTION;
 		}
+		*(type + (ptr - end)) = '\0';
+		ptr = inpstr + matches[RE_REGION_GROUP_NUM].rm_so;
+		errno = 0;
+		*num = strtoul(ptr, NULL, 10);
+		if (errno) {
+			FPGA_MSG("Error parsing number: %s", inpstr);
+			goto out_free;
+		}
+		res = FPGA_OK;
+	} else {
+		res = FPGA_NOT_FOUND;
 	}
+out_free:
+	regfree(&re);
+	return res;
+}
 
-	dir = opendir(device->sysfs_path);
+
+STATIC int find_regions(sysfs_fpga_device *device)
+{
+	int num = -1;
+	char type[8];
+	fpga_result res = FPGA_OK;
+	fpga_result match_res = FPGA_NOT_FOUND;
+	fpga_objtype region_type = FPGA_DEVICE;
+	sysfs_fpga_region **region_ptr = NULL;
+	struct dirent *dirent = NULL;
+	DIR *dir = opendir(device->sysfs_path);
 	if (!dir) {
 		FPGA_MSG("failed to open device path: %s", device->sysfs_path);
-		regfree(&re);
 		return FPGA_EXCEPTION;
 	}
 
 	while ((dirent = readdir(dir)) != NULL) {
+		res = FPGA_OK;
 		if (!strcmp(dirent->d_name, "."))
 			continue;
 		if (!strcmp(dirent->d_name, ".."))
 			continue;
-		reg_res = regexec(&re, dirent->d_name, SYSFS_MAX_RESOURCES,
-				  matches, 0);
-		if (!reg_res) {
-			int type_beg = matches[1].rm_so;
-			// int type_end = matches[1].rm_eo;
-			int num_beg = matches[2].rm_so;
-			// int num_end = matches[2].rm_eo;
-			if (type_beg < 1 || num_beg < 1) {
-				FPGA_MSG("Invalid sysfs resource format");
-				continue;
-			}
-			num = strtoul(dirent->d_name + num_beg, NULL, 10);
-			if (!strncmp(FPGA_SYSFS_FME, dirent->d_name + type_beg,
+
+		match_res = re_match_region(SYSFS_FORMAT(sysfs_region_fmt),
+					    dirent->d_name, type, sizeof(type),
+					    &num);
+		if (match_res == FPGA_OK) {
+			if (!strncmp(FPGA_SYSFS_FME, type,
 				     FPGA_SYSFS_FME_LEN)) {
-				device->fme =
-					make_region(device, dirent->d_name, num,
-						    FPGA_DEVICE);
-			} else if (!strncmp(FPGA_SYSFS_PORT,
-					    dirent->d_name + type_beg,
+				region_type = FPGA_DEVICE;
+				region_ptr = &device->fme;
+			} else if (!strncmp(FPGA_SYSFS_PORT, type,
 					    FPGA_SYSFS_PORT_LEN)) {
-				device->port =
-					make_region(device, dirent->d_name, num,
-						    FPGA_ACCELERATOR);
+				region_type = FPGA_ACCELERATOR;
+				region_ptr = &device->port;
 			}
+			*region_ptr = make_region(device, dirent->d_name, num,
```

Comment:
Are we absolutely certain that region_ptr points to the right thing here?  What if both strncmp() calls fail?

Reply:
Good call

Reply:
Actually, we shouldn't see anything other than fme|port because of the regex pattern.